File: src/checkers-cpp/StudentAI.cpp
#include "StudentAI.h"


//The following part should be completed by students.
//The students can modify anything except the class name and exisiting functions and varibles.
StudentAI::StudentAI(int col,int row,int p)
	:AI(col, row, p)
{
    board = Board(col,row,p);
    board.initializeGame();
    player = WHITE;
    root = new MCTSNode(player, board);
}

StudentAI::~StudentAI() {
    MCTSNode::DeleteTree(root);
}

Move StudentAI::GetMove(Move move) {
    if (move.seq.empty()) {
        player = BLACK;
        delete root;
        root = new MCTSNode(player, board);

    } else {
        board.makeMove(move, player == WHITE ? BLACK : WHITE);
        
        MCTSNode* node = root->SelectChild(move);

        if(node == nullptr) {
            node = new MCTSNode(player, board, move, nullptr);
            MCTSNode::DeleteTree(root);
        }
        else {
            for(auto child : root->children) {
                if(child != node) {
                    MCTSNode::DeleteTree(child);
                }
            }
            root->children.clear();

            delete root;
        }

        root = node;
        node->parent = nullptr;
    }

    Move bestMove = RunMCTS();
    board.makeMove(bestMove, player);
    return bestMove;
}


Move StudentAI::RunMCTS() {
    std::srand(std::time(0));

    // Run MCTS for a fixed number of iterations
    for (int i = 0; i < MCTS_ITERATIONS; i++) {
        // Selection
        MCTSNode* node = root;
        while (node->isFullyExpanded()) {
            auto next = node->SelectChild();
            if(next == nullptr) {
                // Found a terminal state
                int result = node->board.isWin(node->player);
                backpropagate(node, result);
                break;
            }

            node = next;
        }

        // Expansion
        const int newPlayer = node->player == BLACK ? WHITE : BLACK;
        int winner = node->board.isWin(newPlayer);
        if (winner == 0) {
            Board newBoard = node->ExpandNode();
            winner = Rollout(newBoard, node->player);
        }

        // Backpropagation
        backpropagate(node, winner);
    }

    MCTSNode* bestChild = root->SelectBestChild();

    for(auto child : root->children) {
        if(child != bestChild) {
            MCTSNode::DeleteTree(child);
        }
    }
    delete root;
    root = bestChild;
    bestChild->parent = nullptr;
    Move bestMove = bestChild->move;

    return bestMove;
}

int StudentAI::Rollout(Board& board, int prevPlayer) const {
    // maybe do early stopping after a certain number of moves its just whoever has more material or a tie
    while (true) {
        int winner = board.isWin(prevPlayer);
        if (winner != 0) {
            return winner;
        }

        prevPlayer = prevPlayer == 1 ? 2 : 1;
        auto moves = board.getAllPossibleMoves(prevPlayer);

        // Random move selection
        int i = rand() % moves.size();
        int j = rand() % moves[i].size();
        board.makeMove(moves[i][j], prevPlayer);

    }
}

// taking forever
void StudentAI::backpropagate(MCTSNode* node, int winner) {
    while (node != nullptr) {
        node->visits++;
        // if (winner == player || winner == -1) {
        if (winner == player) {
            node->wins++;
        }
        else if (winner == -1) {
            node->wins += 0.5;
        }

        node = node->parent;
    }
}




MCTSNode::MCTSNode(int player, Board board, Move move, MCTSNode* parent)
    : player(player), move(move), board(board), parent(parent), visits(0), wins(0) {
    numPossibleMovesLeft = 0;
    for(const auto& moves : board.getAllPossibleMoves(player)) {
        for(const auto& m : moves) {
            numPossibleMovesLeft++;
        }
    }
}

void MCTSNode::DeleteTree(MCTSNode* root) {
    if(root == nullptr) {
        return;
    }

    for (auto child : root->children) {
        DeleteTree(child);
    }
    delete root;
}

double MCTSNode::UCTValue(double c) const {
    if (visits == 0) return std::numeric_limits<double>::infinity();
    return (static_cast<double>(wins) / visits) + c * std::sqrt(std::log(parent->visits) / visits);
}

MCTSNode* MCTSNode::SelectChild() const {
    MCTSNode* bestChild = nullptr;
    double bestValue;
    bestValue = -std::numeric_limits<double>::infinity();

    for (const auto& child : children) {
        double uctValue = child->UCTValue();
        if (uctValue > bestValue) {
            bestValue = uctValue;
            bestChild = child;
        }
    }
    return bestChild;
}

MCTSNode* MCTSNode::SelectChild(const Move& move) const {
    for (const auto& child : children) {
        if(child->move.seq.size() != move.seq.size()) {
            continue;
        }

        bool same = true;
        for(int i = 0; i < move.seq.size(); i++) {
            if(child->move.seq[i].x != move.seq[i].x || child->move.seq[i].y != move.seq[i].y) {
                same = false;
            }
        }

        if(same) {
            return child;
        }
    }

    return nullptr;
}

MCTSNode* MCTSNode::SelectBestChild() const {
    MCTSNode* bestChild = nullptr;
    int bestValue = 0;
    for (const auto& child : children) {
        if (child->visits > bestValue) {
            bestValue = child->visits;
            bestChild = child;
        }
    }
    return bestChild;
}

int evaluateMaterial(vector<vector<Checker>>& board) {
    int material = 0;
    int BOARD_SIZE = board.size();
    for (int i = 0; i < BOARD_SIZE; ++i) {
        for (int j = 0; j < BOARD_SIZE; ++j) {
            if (board[i][j].color == "W" && !board[i][j].isKing) {
                material -= 1;
            } else if (board[i][j].color == "W" && board[i][j].isKing) {
                material -= 3;
            } else if (board[i][j].color == "B" && !board[i][j].isKing) {
                material += 1;
            } else if (board[i][j].color == "B" && board[i][j].isKing) {
                material += 3;
            }
        }
    }
    return material;
}

int evaluateCenterControl(vector<vector<Checker>>& board) {
    int eval = 0;
    int BOARD_SIZE = board.size();

    if (BOARD_SIZE < 5) {
        return eval;
    }

    for (int i = BOARD_SIZE/2 - 1; i < BOARD_SIZE/2 + 1 + (BOARD_SIZE % 2); ++i) {
        for (int j = BOARD_SIZE/2 - 1; j < BOARD_SIZE/2 + 1 + (BOARD_SIZE % 2); ++j) {
            if (board[i][j].color == "W" && !board[i][j].isKing) {
                eval -= 1;
            } else if (board[i][j].color == "W" && board[i][j].isKing) {
                eval -= 3;
            } else if (board[i][j].color == "B" && !board[i][j].isKing) {
                eval += 1;
            } else if (board[i][j].color == "B" && board[i][j].isKing) {
                eval += 3;
            }
        }
    }

    return eval;
}

double evaluateProgression(vector<vector<Checker>>& board) {

    double avgW = 0;
    double avgB = 0;

    int numW = 0;
    int numB = 0;

    int eval = 0;
    int BOARD_SIZE = board.size();
    for (int i = 0; i < BOARD_SIZE; ++i) {
        for (int j = 0; j < BOARD_SIZE; ++j) {
            if (board[i][j].color == "W" && !board[i][j].isKing) {
                avgW += BOARD_SIZE-1-i;
                numW += 1;
            } else if (board[i][j].color == "B" && !board[i][j].isKing) {
                avgB += i;
                numB += 1;
            }
        }
    }

    if (numW == 0 || numB == 0) {
        return eval;
    }

    avgW /= static_cast<double>(numW);
    avgB /= static_cast<double>(numB);

    return avgB - avgW;
}

// Heuristic function to evaluate the current board state
double evaluateBoard(vector<vector<Checker>>& board) {
    double score = 0;

    score += static_cast<double>(evaluateMaterial(board));
    score += static_cast<double>(evaluateCenterControl(board));
    score += 7*evaluateProgression(board);

    return score;
}

void MCTSNode::createChildren() {
    for(auto& moves : board.getAllPossibleMoves(player)) {
        for(auto& m : moves) {
            Board newBoard = CopyBoard(board);
            newBoard.makeMove(m, player);

            MCTSNode *child = new MCTSNode(player == BLACK ? WHITE : BLACK, newBoard, m, this);
            children.push_back(child);
        }
    }
}

Board MCTSNode::ExpandNode() {
    if(children.empty()) {
        createChildren();
    }


    double bestScore;
    const int newPlayer = player == BLACK ? WHITE : BLACK;

    std::function<double(double, double)> keepBestScore;
    if(player == BLACK) {
        keepBestScore = [](double a, double b) { return std::max(a, b); };
        bestScore = -std::numeric_limits<double>::infinity();
    }
    else if(player == WHITE) {
        keepBestScore = [](double a, double b) { return std::min(a, b); };
        bestScore = std::numeric_limits<double>::infinity();
    }

    int bestMoveIndex = 0;
    int i = 0;
    for (; i < children.size(); i++) {
        MCTSNode* child = children[i];
        if(child->visits > 0) {
            continue;
        }

        // some heuristic
        double score = evaluateBoard(child->board.board);
        double newBestScore = keepBestScore(bestScore, score);

        if(newBestScore != bestScore) {
            bestScore = newBestScore;
            bestMoveIndex = i;
        }
    }

    Board newBoard = children[bestMoveIndex]->board;
    numPossibleMovesLeft--;
    return newBoard;
}


Board MCTSNode::CopyBoard(const Board& board) {
    Board newBoard;
    vector<vector<Checker>> boardVec;
    for(int i = 0; i < board.board.size(); i++) {
        vector<Checker> row;
        for(int j = 0; j < board.board[i].size(); j++) {
            auto checker = Checker(board.board[i][j].color, i, j);
            checker.isKing = board.board[i][j].isKing;
            row.push_back(checker);
        }
        boardVec.push_back(row);
    }
    newBoard.col = board.col;
    newBoard.row = board.row;
    newBoard.p = board.p;
    newBoard.blackCount = board.blackCount;
    newBoard.whiteCount = board.whiteCount;
    newBoard.tieCount = board.tieCount;
    newBoard.tieMax = board.tieMax;
    newBoard.board = boardVec;

    return newBoard;
}


inline bool MCTSNode::isFullyExpanded() {
    return numPossibleMovesLeft == 0;
}



// void MCTSNode::showTree(MCTSNode* root, int level) {
//     MCTSNode* current = root;
//     if(current->visits > 0) {
//         cout << std::setw(level) << "" << " Wins: " << current->wins << " Player: " << current->player << " Move: " << current->move.toString() << endl;
//     }
//
//     for(auto child : current->children) {
//         if(level < 2) {
//             showTree(child, level + 1);
//         }
//     }
// }

File: src/checkers-python/StudentAI.py
from random import randint
from BoardClasses import Move
from BoardClasses import Board
from Checker import Checker
import math

def copyBoard(oldboard):
        newBoard = Board(oldboard.col, oldboard.row, oldboard.p)
        newBoard.board = [[0 for _ in newBoard.col] for _ in newBoard.row]
        for i in range(len(newBoard.board)):
            for j in range(len(newBoard.board[0])):
                newBoard.board[i][j] = Checker(oldboard.board[i][j].color, i, j)
                newBoard.board[i][j].is_king = oldboard.board[i][j].is_king
        newBoard.col = oldboard.col
        newBoard.row = oldboard.row
        newBoard.p = oldboard.p
        newBoard.black_count = oldboard.black_count
        newBoard.tie_counter = oldboard.tie_counter
        return newBoard

class MCTSNode:

    def __init__(self, player, move=0, parent=None):
        self.player = player
        self.move = move
        self.parent = parent
        self.children = []

        self.visits = 0
        self.wins = 0

    def UCTValue(self, constant):
        if not visits: return float('inf')
        return (wins/visits) + c * math.sqrt(math.log(self.parent.visits)/visits)
    
    def SelectChild(self):
        bestValue = float('-inf')
        bestChild = None

        for child in children:
            uct = child.UCTValue()
            if (uct > bestValue):
                bestValue = uct
                bestChild = child
        return bestChild
    
    def evaluateMaterial(self, board):
        material = 0
        for i in range(len(board)):
            for j in range(len(board[0])):
                piece = board[i][j]
                if piece.is_king:
                    if piece.color == 'W':
                        material += 3
                    elif piece.color == 'B':
                        material -= 3
                else:
                    if piece.color == 'W':
                        material += 1
                    elif piece.color == 'B':
                        material -= 1
        return material
    
    def evaluateBoard(self, board):
        return float(evaluateMaterial(board))
    
    

    def bestMove(self, board, possible_moves):
        bestMove = Move(possible_moves)
        bestScore = 0

        def keepBestScore(a, b):
            return max(a,b)

        bestScore = float('-inf')
        
        newPlayer = 2 if player == 1 else 1

        for move_set in possible_moves:
            for m in move_set:
                # Try the move
                newBoard = copyBoard(board)
                newBoard.make_move(m, player)

                newNode = MCTSNode(newPlayer, m, self)
                children.append(newNode)

                # Apply heuristic
                score = evaluateBoard(newBoard.board)
                newBestScore = keepBestScore(bestScore, score)

                if newBestScore != bestScore:
                    bestScore = newBestScore
                    bestMove = m
        
        return bestMove



#The following part should be completed by students.
#Students can modify anything except the class name and exisiting functions and varibles.
class StudentAI():

    def __init__(self,col,row,p):
        self.col = col
        self.row = row
        self.p = p
        self.board = Board(col,row,p)
        self.board.initialize_game()
        self.color = ''
        self.opponent = {1:2,2:1}
        self.color = 2

        self.player = 2
        self.MCTS_ITERATIONS = 1000
    
    def get_move(self,move):
        if len(move) != 0:
            self.board.make_move(move,self.opponent[self.color])
        else:
            self.color = 1
        moves = self.board.get_all_possible_moves(self.color)
        index = randint(0,len(moves)-1)
        inner_index =  randint(0,len(moves[index])-1)
        move = moves[index][inner_index]
        self.board.make_move(move,self.color)
        return move



    def RunMCTS(self):
        root = MCTSNode

        for i in range(MCTS_ITERATIONS):

            # Select
            node = root
            newBoard = copyBoard(self.board)

            while (node.visits > 0):
                prevPlayer = node.player
                next_node = node.SelectChild()
                if not next_node:
                    result = newBoard.is_win(prevPlayer)
                    self.backpropagate(node, result)
                    break
                
                node = next_node
                newBoard.make_move(node.move, prevPlayer)

            # Expand
            possible_moves = newBoard.get_all_possible_moves(node.player)
            winner = 0
            if len(possible_moves) == 0:
                winner = newBoard.is_win(2 if node.player == 1 else 1)
            else:
                bestMove = node.bestMove(newBoard, possible_moves)
                newBoard.make_move(bestMove, node.player)

                winner = self.rollout(newBoard, node.player)

            # Backpropagate
            self.backpropagate(node, winner)
    
    def rollout(self, board, prevPlayer):
        while True:
            winner = board.is_win(prevPlayer)
            if winner:
                return winner

            prevPlayer = 2 if prevPlayer == 1 else 1
            moves = board.get_all_possible_moves(prevPlayer)
            
            
            i = randint(0, len(moves)-1)
            j = randint(0, len(moves[i])-1)
            board.make_move(moves[i][j], prevPlayer)


    def backpropagate(self, node, result):
        while node:
            node.visits += 1
            if winner == node.player or winner == -1:
                node.wins += 1
            node = node.parent
